zips/zip-0231.md

```diff
@@ -0,0 +1,164 @@
+
+In the light client protocol, a recipient need not download full transaction
+information if the compact transaction information indicates that they do not
+receive any memo in the transaction.
+
+
+# Requirements
+
+- Recipients can receive memo data that is greater than 512 bytes in length.
+- Multiple recipients, across any of the shielded pools, can be given the
+  capability to view the same memo data.
+- The exact number and exact lengths of distinct decryptable memos should not
+  be revealed, even to the transaction recipients, although an upper bound on
+  the total length of memo data that the observer does not have the capability
+  to view will be leaked.
+- A recipient can determine whether or not they have been given the capability
+  to view any memo solely by decrypting the note ciphertext.
+- Memo chunks within a transaction can be individually pruned without preventing
+  the transaction from being verified.
+  - As a consequence, a recipient with the capability to view a memo can prune
+    all of that memo's chunks.
+
+
+# Non-requirements
+
+- Recipients do not need to be able to receive multiple memos per note.
+
+
+# Specification
+
+## Memo encryption
+
+Sketch:
+
+In shielded bundles, the 512-byte memo field is replaced by `[flag_byte] || memo_key`.
+
+```
+STREAM: AEAD(key, counter, pt)
+
+[
+    (key_a, 0),
+    (key_b, 0),
+    (key_a, 1),
+    (key_c, 0),
+    (key_c, 1),
+    (key_a, 2),
+]
+```
+
+## Memo bundle format
+
+```rust
+enum MemoChunk {
+    Encrypted(AEAD(MemoChunk, memo_key)),
+    Pruned {
+        memo_chunk_digest: Digest,
+    }
+}
+
+struct MemoBundle {
+    chunks: Vec<MemoChunk>,
+}
+```
+
+## Encoding in transactions
+
+TODO: Decide whether / how to enable memo data to be pruned from a transaction while still enabling it to be broadcast.
+
+```rust
+enum MemoBundle {
+    Chunks(Vec<MemoChunk>),
+    Pruned {
+        memo_bundle_digest: Digest,
+    }
+}
+```
+
+## Transaction sighash
+
+```
+memo_chunk_digest = H(AEAD(MemoChunk, memo_key))
+memo_bundle_digest = H(concat(memo_chunk_digests))
+```
+
+The memo bundle digest structure is a performance optimisation for the case
+where all memo chunks in a transaction have been pruned.
+
+
+# Rationale
+
+## Pruned encoding
+
+The separation of memo data from note data, and the new ability to easily store
+variable-length memo data, opens up an attack vector against node operators for
+storing arbitrary data. The transaction digest commitments to the memo bundle
+are structured such that if a node operator is presented with a memo key (i.e.
+they are given the capability to decrypt a particular memo), they can identify
+and prune the corresponding memo chunks, while still enabling the transaction to
+be validated as part of its corresponding block and broadcast over the network.
+
+Note that broadcasting a partially-pruned transaction means that the pruned
+chunks no longer contribute to the upper bound on memo data.
```

Comment:
We discussed fees, and decided to propose a simple modification to ZIP 317 where a memo bundle may contain one free chunk if there are any shielded outputs in the transaction, and otherwise each memo chunk requires `marginal_fee` (the same 5000 zatoshis as for logical actions). This has the following properties:
- The required fee to add more memo chunks scales at the same rate as adding logical actions, so it isn't a cheaper mechanism for an adversary to bloat chain size.
- The "baseline transaction" (one spent note, one output to an external recipient with a memo, one change output without a memo) is the same fee as before.
- The "broadcast transaction" (many outputs to different recipients all given the same memo) is the same fee as before (but a smaller transaction).
- The "many memos transaction" (many outputs to different recipients all with unique memos) is at most around twice the fee as before.

Reply:
Additionally, the "many memos transaction" then is essentially charged the same fee as sending a separate transaction to each recipient and taking advantage of the free memo chunk (because the base fee is double the cost of adding a recipient to the transaction.)